include/cuco/bloom_filter.cuh

```diff
@@ -0,0 +1,393 @@
+   *
+   * @tparam InputIt Device-accessible random access input iterator where
+   * <tt>std::is_convertible<std::iterator_traits<InputIt>::value_type,
+   * bloom_filter<K>::key_type></tt> is `true`
+   * @tparam StencilIt Device-accessible random-access iterator whose `value_type` is
+   * convertible to Predicate's argument type
+   * @tparam Predicate Unary predicate callable whose return type must be convertible to `bool` and
+   * argument type is convertible from <tt>std::iterator_traits<StencilIt>::value_type</tt>
+   * @tparam OutputIt Device-accessible output iterator assignable from `bool`
+   *
+   * @param first Beginning of the sequence of keys
+   * @param last End of the sequence of keys
+   * @param stencil Beginning of the stencil sequence
+   * @param pred Predicate to test on every element in the range `[stencil, stencil +
+   * std::distance(first, last))`
+   * @param output_begin Beginning of the sequence of booleans for the presence of each key
+   * @param stream CUDA stream this operation is executed in
+   */
+  template <class InputIt, class StencilIt, class Predicate, class OutputIt>
+  __host__ void test_if_async(InputIt first,
+                              InputIt last,
+                              StencilIt stencil,
+                              Predicate pred,
+                              OutputIt output_begin,
+                              cuda::stream_ref stream = {}) const noexcept;
+
+  /**
+   * @brief Gets a pointer to the underlying filter storage.
+   *
+   * @return Pointer to the underlying filter storage
+   */
+  [[nodiscard]] __host__ word_type* data() noexcept;
+
+  /**
+   * @brief Gets a pointer to the underlying filter storage.
+   *
+   * @return Pointer to the underlying filter storage
+   */
+  [[nodiscard]] __host__ word_type const* data() const noexcept;
+
+  /**
+   * @brief Gets the number of sub-filter blocks.
+   *
+   * @return Number of sub-filter blocks
+   */
+  [[nodiscard]] __host__ extent_type block_extent() const noexcept;
+
+  /**
+   * @brief Gets the number of bits in a key's fingerprint
+   *
+   * @return The number of fingerprint bits
+   */
+  [[nodiscard]] __host__ hasher hash_function() const noexcept;
+
+  /**
+   * @brief Gets the function used to hash keys
+   *
+   * @return The function used to hash keys
+   */
+  [[nodiscard]] __host__ std::uint32_t pattern_bits() const noexcept;
+
+  /**
+   * @brief Gets the allocator.
+   *
+   * @return The allocator
+   */
+  [[nodiscard]] __host__ allocator_type allocator() const noexcept;
+
+  /**
+   * @brief Get device ref.
+   *
+   * @return Device ref of the current `bloom_filter` object
+   */
+  [[nodiscard]] __host__ ref_type<> ref() const noexcept;
+
+ private:
+  allocator_type allocator_;  ///< Allocator used to allocate device-accessible storage
+  std::unique_ptr<word_type, detail::custom_deleter<std::size_t, allocator_type>>
+    data_;          ///< Storage of the current `bloom_filter` object
```

Comment:
What makes us drop `aow_storage`?

Reply:
I found `aow_storage` too rigid. For one, we want to allow users to serialize the underlying filter storage. It's straightforward to serialize a vector of integers or even bytes compared to a more complex type such as `cuda::std::array`. The second problem that I ran into is that I sometimes need to perform a (vectorized) load of only part of a window, e.g., if the filter has 16*4byte words and I want to perform a lookup with a cooperative group of size 2. If I split the block among the two threads both can emit an efficient `LDG.128` rather than a single thread loading both sectors.